core/drivers/nxp_ls_i2c.c

```diff
@@ -0,0 +1,420 @@
+		return TEE_ERROR_BUSY;
+
+	if (test_rx_ack && (reg & I2C_IBSR_RXAK))
+		return TEE_ERROR_NO_DATA;
+
+	if (reg & I2C_IBSR_TCF)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_read(struct i2c_regs *regs, uint32_t slave_address,
+			   struct i2c_operation *operation,
+			   bool is_last_operation)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	uint32_t index = 0;
+
+	// Write Slave Address
+	io_write8((vaddr_t)&regs->ibdr, (slave_address << BIT(0)) | BIT(0));
+	status = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (status != TEE_SUCCESS)
+		return status;
+
+	// select Receive mode.
+	i2c_io_and8((vaddr_t)&regs->ibcr, ~I2C_IBCR_TXRX);
+	if (operation->length_in_bytes > 1) {
+		// Set No ACK = 0
+		i2c_io_and8((vaddr_t)&regs->ibcr, ~I2C_IBCR_NOACK);
+	}
+
+	// Perform a dummy read to initiate the receive operation.
+	io_read8((vaddr_t)&regs->ibdr);
+
+	for (index = 0; index < operation->length_in_bytes; index++) {
+		status = i2c_transfer_complete(regs, I2C_BUS_NO_TEST_RX_ACK);
+		if (status != TEE_SUCCESS)
+			return status;
+		if (index == (operation->length_in_bytes - 2)) {
+			// Set No ACK = 1
+			i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_NOACK);
+		} else if (index == (operation->length_in_bytes - 1)) {
+			if (!is_last_operation) {
+				// select Transmit mode (for repeat start)
+				i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_TXRX);
+			} else {
+				// Generate Stop Signal
+				i2c_io_and8((vaddr_t)&regs->ibcr,
+					    ~(I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+				status = i2c_bus_test_bus_busy
+					(regs, I2C_BUS_TEST_IDLE);
+				if (status != TEE_SUCCESS)
+					return status;
+			}
+		}
+		operation->buffer[index] = io_read8((vaddr_t)&regs->ibdr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_write(struct i2c_regs *regs, uint32_t slave_address,
+			    struct i2c_operation *operation)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	uint32_t index = 0;
+
+	// Write Slave Address
+	io_write8((vaddr_t)&regs->ibdr,
+		  (slave_address << BIT(0)) & (uint8_t)(~(BIT(0))));
+	status = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (status != TEE_SUCCESS)
+		return status;
+
+	// Write Data
+	for (index = 0; index < operation->length_in_bytes; index++) {
+		io_write8((vaddr_t)&regs->ibdr, operation->buffer[index]);
+		status = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+		if (status != TEE_SUCCESS)
+			return status;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_stop(struct i2c_regs *regs)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	uint8_t reg = 0;
+
+	status = TEE_SUCCESS;
+	reg = io_read8((vaddr_t)&regs->ibsr);
+	if (reg & I2C_IBSR_IBB) {
+		// Generate Stop Signal
+		i2c_io_and8((vaddr_t)&regs->ibcr,
+			    ~(I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+		status = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_IDLE);
+		if (status != TEE_SUCCESS)
+			return status;
+	}
+
+	// Disable I2c Controller
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+
+	return status;
```

Comment:
We can only reach this point if `status` is `TEE_SUCCESS`. By explicitly returning `TEE_SUCCESS` it would be more clear that we're good if we've reached this point. Besides it's a different pattern compared to the functions above.